Align datasource enrichment

Sefi4 opened the pull request:
Before this PR, `localmanager` operator and `kubemanager` operator were missing podLabels and containerStartedAt enrichment.
This PR adds it to have image-based gadget enrichment aligned with built-in

## How to use

[ describe what reviewers need to do in order to validate this PR ]

## Testing done
Consuming the localmanager operator locally and validating the output:
![image](https://github.com/user-attachments/assets/6f745784-ad85-496d-84ba-2122aa1bd7db)

mqasimsarfraz commented on pkg/datasource/compat/wrapper.go:
```diff
@@ -358,6 +379,10 @@ func (ev *EventWrapper) SetPodMetadata(container types.Container) {
 			ev.ownerNameAccessor.PutString(ev.Data, container.K8sOwnerReference().Name)
 		}
+		if ev.podLabelsAccessor.IsRequested() {
+			podLabels, _ := json.Marshal(k8s.PodLabels)
```
I am thinking if it will make sense combine the labels as string till we support https://github.com/inspektor-gadget/inspektor-gadget/issues/3032 @alban / @mauriciovasquezbernal thoughts?

burak-ok replied:
I would suggest the same. For the time I would keep it the same as the `kubectl ... --show-labels` output:

```
➭ kubectl get all -A --show-labels
NAMESPACE     NAME              READY   STATUS   RESTARTS    AGE    LABELS
gadget        pod/gadget-qbfhr  1/1     Running  0           6h2m   controller-revision-hash=5b9ccb6798,k8s-app=gadget,pod-template-generation=1
```

So it would be `key1=value1,key2=value2`

burak-ok commented:
In the commit message there are many lines from `dependabot`, which shouldn't be there.

Additionally the addition to `trace_exec` should be in a separate commit, even better a separate PR

Sefi4 commented:
> In the commit message there are many lines from `dependabot`, which shouldn't be there.
> 
> Additionally the addition to `trace_exec` should be in a separate commit, even better a separate PR
Fixed
https://github.com/inspektor-gadget/inspektor-gadget/pull/3695

burak-ok reviewed (approved):
LGTM, thanks

burak-ok commented:
So the tests are currently failing because they want to unmarshall the `podLabels` string back to `map[string]string`

Either of the options is needed:
1. Datasource support for array types: https://github.com/inspektor-gadget/inspektor-gadget/issues/3032
2. Adjust the function that does the unmarshalling for `podLabels` or arrays in general

Sefi4 commented on pkg/operators/filter/filter.go:
```diff
@@ -426,3 +426,5 @@ func getFilterFunc(f datasource.FieldAccessor, op comparisonType, negate bool, s
 	operators.RegisterDataOperator(&filterOperator{})
 }
+
+var FilterOperator = &filterOperator{}
```
For users which consume IG as a Golang library, and willing to use the filter operator, they would need to pass a pointer to the operator to the `gadgetcontext.WithDataOperators` option. To enable this behavior, we need to export the filter operator in some way. I chose global variable, since ocihanelder and clioperator use this way too, so I wanted to stay consistent with the coding style of the go library from the consumers standpoint.

Sefi4 replied:
Done

Sefi4 commented on pkg/datasource/compat/wrapper.go:
```diff
@@ -299,6 +312,14 @@ func WrapAccessors(source datasource.DataSource, mntnsidAccessor datasource.Fiel
+	ev.containerStartedAtAccessor, err = runtime.AddSubField(
+		"containerStartedAt",
+		api.Kind_String,
+		datasource.WithOrder(-21),
```
Yes! I'll change it

Sefi4 replied:
Yeap, done :)

Sefi4 commented:
> So the tests are currently failing because they want to unmarshall the `podLabels` string back to `map[string]string`
> 
> Either of the options is needed:
> 
> 1. Datasource support for array types: [datasource: Support array fields #3032](https://github.com/inspektor-gadget/inspektor-gadget/issues/3032)
> 2. Adjust the function that does the unmarshalling for `podLabels` or arrays in general


Chose second option, also favored by @eiffel-fl. Added a custom Unmrashl function to deal with `K8sMetadata` type. Also added a test to `match_test.go` to validate correct decode by tests.

Sefi4 commented on pkg/types/utils.go:
```diff
@@ -0,0 +1,16 @@
+package types
```
Yea I was wondering about this too

Sefi4 commented on pkg/types/types.go:
```diff
@@ -428,3 +445,19 @@ type WithNetNsID struct {
+		if len(kv) == 2 {
+			labels[kv[0]] = kv[1]
+		} else {
+			err = fmt.Errorf("invalid pod labels format. Format should be key=value")
```
@eiffel-fl - by explode you mean panic? I thought that if we can avoid it then we should.
Regarding the suggested change, if there are several pod labels in the wrong format the error would only show the last invalid pair. We can overcome this by appending all invalid pairs, but possibly that could get quite big error which I thought it's better to avoid.

TBH out of all options I think current implementation is best, but I'll do as you'll decide :)

Sefi4 commented on pkg/testing/match/match_test.go:
```diff
@@ -21,13 +21,34 @@ import (
 }
 
+func TestPodLabelsUnmarshal(t *testing.T) {
+	input := `{"node": "node1", "namespace": "default", "podName": "test", "podLabels": "app=nginx,foo=bar"}`
+	expected := &types.K8sMetadata{
+		Node: "node1",
+		BasicK8sMetadata: types.BasicK8sMetadata{
+			Namespace: "default",
+			PodName:   "test",
+			PodLabels: map[string]string{
+				"app": "nginx",
+				"foo": "bar",
+			},
+		},
+	}
+	actual := decodeJSONOutput(t, JSONSingleObjectMode, input, func(*types.K8sMetadata) {})
+
+	require.Len(t, actual, 1)
+	assert.Equal(t, expected, actual[0])
```
Good idea, I thought it would be better to test the custom Unmarshal function itself, WDYT?

Sefi4 commented:
> > Now tests are failing due to permissions: ![image](https://private-user-images.githubusercontent.com/62463910/388013802-501804da-7619-4075-a97e-9e558fd66144.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIwOTgyOTksIm5iZiI6MTczMjA5Nzk5OSwicGF0aCI6Ii82MjQ2MzkxMC8zODgwMTM4MDItNTAxODA0ZGEtNzYxOS00MDc1LWE5N2UtOWU1NThmZDY2MTQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIwVDEwMTk1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBlNGEyYTM4YzRhMzI4ZmZiNmJlNWMxYmJhMDIxZmNlNzBiOTcwYmZkYjAwZTE5NjVkNDJkYmNlY2Y0Y2RhMGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.QspXy6_NiVPw_4R_mn7nypNCrZdTEkIvzpNjSlOU7Hk)
> 
> I do not know the root cause for this, let's run it again and see whether this is due to flakyness. I would rather focus on this one though:
> 
> > 2024-11-20T08:28:32.5735363Z === RUN   TestLoad
> > 2024-11-20T08:28:32.5735772Z     advisor_test.go:35: parsing line 1: invalid pod labels format: expected key=value, got
> > 2024-11-20T08:28:32.5735911Z --- FAIL: TestLoad (0.00s)

Ahh I missed that failure in the logs. Fixed that, by skipping empty label instead of erroring on it.

Sefi4 commented on pkg/types/types.go:
```diff
@@ -191,6 +193,41 @@ type CommonData struct {
+	case map[string]interface{}:
+		k.PodLabels = make(map[string]string)
+		for key, value := range v {
+			if strVal, ok := value.(string); ok {
+				k.PodLabels[key] = strVal
+			} else {
+				return fmt.Errorf("invalid value type for key %s in PodLabels", key)
+			}
```
We can't use strVal outside the scope of the if statement. To remove the need in else statement we need to do something like this:
![image](https://github.com/user-attachments/assets/d9cf26c1-fe6a-404d-89ab-d380a604734f)

I like the current styling better, but that's just a matter of taste :) WDYT?

Sefi4 commented on pkg/operators/filter/filter.go:
```diff
@@ -426,3 +426,5 @@ func getFilterFunc(f datasource.FieldAccessor, op comparisonType, negate bool, s
 	operators.RegisterDataOperator(&filterOperator{})
 }
+
+var FilterOperator = &filterOperator{}
```
Let's give it one more try please and if it doesn't pass the tests, I'll move it to another PR

Sefi4 replied:
Moved the filter operator exporting to another PR: https://github.com/inspektor-gadget/inspektor-gadget/pull/3774

Sefi4 commented:
All the tests' issues are because the unmrashal function changes the zero value of map[string]string (podLabels) to map[string]string{} as opposed to nil. I tried to keep the unmrashal function consistent with the native zero function hope it would work.

I can't really test it locally because minikube is broken on my machine for some reason

Sefi4 commented:
> Let's see how it goes with the CI.

Ah, I don't think they will pass, I didn't change anything except for reverting the commit for exporting Filter operator

Sefi4 commented on examples/go.mod:
```diff
@@ -14,7 +14,7 @@ require (
 
 require (
-	github.com/AdaLogics/go-fuzz-headers v0.0.0-20230811130428-ced1acdcaa24 // indirect
+	github.com/AdaLogics/go-fuzz-headers v0.0.0-20240806141605-e8a1dd7889d6 // indirect
```
The CI tests failed on the examples dir not having up to date go.mod

Sefi4 replied:
Should I try to revert it, and let's see if the CI/CD tests pass?

Sefi4 replied:
Done @eiffel-fl

Sefi4 replied:
It added some changed to the DNS gadget, though I didn't do any change related to it

Sefi4 commented:
Hey @eiffel-fl, finally the tests seem to pass! Can we merge it?

Sefi4 commented on pkg/gadgets/trace/dns/tracer/dnswithlongpaths_bpfel.go:
```diff
@@ -127,8 +127,9 @@ type dnsWithLongPathsMapSpecs struct {
-	PortsLen *ebpf.VariableSpec `ebpf:"ports_len"`
+	Ports       *ebpf.VariableSpec `ebpf:"ports"`
+	PortsLen    *ebpf.VariableSpec `ebpf:"ports_len"`
+	Unusedevent *ebpf.VariableSpec `ebpf:"unusedevent"`
```
Awesome, thanks

Sefi4 replied:
Thanks, done
